### ISSUE TITLE
introduce variant_view

### ISSUE THREAD

**Issue by mgaunard** [ericniebler/range-v3#425]
It would be nice to have a variant_view, like any_view but limited to a closed set of ranges.
I think this could be useful for flattening trees, possibly not as bad as any_view.

**Comment by gnzlbg** [comment]
What API would you want that to have? / What would you like that to do?

Right now if you have 

```c++ 
std::vector<A> as;
std::vector<B> bs;
```

you already can map those to a variant type

```c++
as_v = as | view::transform([](auto&& a) { return variant<A,B,C>{a}; });
bs_v = bs | view::transform([](auto&& b) { return variant<A,B,C>{b}; });
```

and then do anything to that, like concat them, process them:

```c++
auto rng = view::concat(as, bs) | view::transform([](auto&& v_) {
  return visit([](auto&& v) {
    using v_t = uncvref_t<decltype(v)>;
    if constexpr(Same<v_t, A>{}) {
      // do something with v
      return v_t{v};
    else if constexpr(Same<v_t, B>{}) {
      // do something else with v
      return v_t{v};
    } else {  // C
      return v_t{v};
    }, v_);
});
```

and afterwards, filter the range by the variant type and transform them back to ranges of a single type.

**Comment by mgaunard** [comment]
My use case was actually something like this:

```
struct object
{
};

struct archive
{
   vector<object>& objects();
};

struct expr : variant<archive, object>
{
    auto objects()
    {
        if(holds_alternative<object>(*this))
            return view::single(get<object>(*this));
        else
            return get<archive>(*this).objects();
    }
};
```

The two returns have different types, and currently the only way to make that work is with any_view.
Now I guess I could use concat if somehow I could build empty dummy ranges.

**Comment by gnzlbg** [comment]
> Now I guess I could use concat if somehow I could build empty dummy ranges.

There is a `view::empty<object>()` :)

> My use case was actually something like this:

To make `objects()` return you need to return a single range type. Since your range's value_type is a single type `object`, and vector has contiguous iterators I would suggest:  

```c++
auto objects() {  //
       if(holds_alternative<object>(*this)) {
            object& o = get<object>(*this);
            return make_iterator_range(&o, &o + 1);
       } else {
            vector<object>& os = get<archive>(*this).objects();
            return make_iterator_range(os.data(), os.data() + os.size());
      }
}
```

The following could be a solution for the general case. I don't think it currently works, but I think we should make it work:

```c++
auto objects() {  //
      using single_t = decltype(view::single(get<object>(*this));
      using range_t = decltype(get<archive>(*this).objects()); 
      
      optional<single_t> os;
      optional<range_t> or;

      if (holds_alternative<object>(*this)) {
          os = view::single(get<object>(*this);
      } else {
          or = get<archive>(*this).objects();
      }
      
      return view::concat(os, or) | view::flatten;
      }
}
```

Notice that an `optional<X>` is a range (with zero or one elements), and `optional<single_t/range_t>` is a range of ranges of `object`s, so when we `concat`enate them we are concatenating two ranges of ranges of objects, which we then `flatten` into a range of objects.

This doesn't work because `optional` is currently not a range (neither in range-v3 nor in library fundamentals TS), but I think it should be and that:

```c++
for (auto&& x : optional<int>{}) {
   // executed only if optional has a value
}
```

should work as well.

**Comment by gnzlbg** [comment]
The API for a "variant" view that makes this easy could look something like: 

```c++
auto objects() {  //
   return view::variant(
       std::size_t{2},                      // Index
       [&]() { return view::ints(0, 1); },  // void => Range0
       [&]() { return view::single(1); },   // void => Range1
       view::repeat_n(1, 2));               // Range2
}
```

where the user should be able to pass the ranges either directly, or with a function that is called to obtain the range only when the index matches, and where the value types of the ranges should have a common type.

I just don't like the name `variant` for this that much, but the other only name I can think of is `switch_`, which I don't like either.

**Comment by mgaunard** [comment]
I simplified my example by using vector, the real thing is not contiguous.

Instead of having two optionals, and magic to interpret them as ranges, isn't it simpler to just have a variant_range of the two possibilites? (`variant_range<single_t, range_t>`)

To take back the code you wrote

```C++
auto objects() {  //
    using single_t = decltype(view::single(get<object>(*this));
    using range_t = decltype(get<archive>(*this).objects()); 

    return [&]() -> variant_range<single_t, range_t>
    {
        if (holds_alternative<object>(*this)) {
            return view::single(get<object>(*this));
        } else {
            return get<archive>(*this).objects();
        }
    }();
}
```

**Comment by gnzlbg** [comment]
@mgaunard Yes, the optional solution is something that one could quickly use today.

Independently of how the `variant_range` is implemented, I really want a "higher-level" API (like `variant::view` above, that automatically deduces the range types.

**Comment by gnzlbg** [comment]
I think this is a duplicate of #221 .

**Comment by mgaunard** [comment]
aren't interleaved ranges something entirely different?
My understanding is that it's a view that takes two ranges {a0, a1, a2, a3}, {b0, b1, b2, b3} and returns the range {a0, b0, a1, b1, a2, b2, a3, b3}, so it's more like a kind of flattened zip.

**Comment by gnzlbg** [comment]
Yeah it would have been better to say "highly related" and not duplicate. Basically if you want to interleave two ranges whose elements don't have a `common_reference` type, you can either disallow that, use an any, or use a `variant_range`/`variant_iterator`/...

In that issue there is a discussion about `variant_range`/`variant_iterator` that is relevant here as well.
